### ISSUE TITLE
XML Stopped Working

### ISSUE THREAD

**Issue by fraktality** [Dreadbot/Recycle-Rush#3]
XMLInput has stopped feeding input to the drivebase.  We need to either scale it down or fix it.  I've bypassed it so that the input is directly fed to the drivebase class, and everything seems to work in that configuration.
Some bad code was likely pushed to the master branch without being tested.

**Comment by C7C8** [comment]
If you're on the master branch, there's your problem. The latest (tested, working) code is on the develop branch. 

Also, note that when you switch branches to put code from Master on the robot, you need to switch out the XML config file on the robot as well, which is NOT done automatically. The code previously on Master is radically different than our current setup. I am merging Develop into Master to remedy this problem immediately. The code currently set up to run - basic intake, driving, etc. - is stable.

**Comment by C7C8** [comment]
Oh, one more thing - note that we changed the XML config file to use slot 0 for the joystick... you might have just had your joystick in another slot instead of slot 0 on the driverstation USB config.

**Comment by SteveKing** [comment]
On 2015-02-10 09:58 , Sourec wrote:

> Oh, one more thing - note that we changed the XML config file to use 
> slot 0 for the joystick... you might have just had your joystick in 
> another slot instead of slot 0 on the driverstation USB config.

A few points:

1) We're at the point not where we NEED WORKING CODE. At all times. 
Right now the "Stable_V1" tag isn't stable. Our top priority today has 
to be getting a stable, drivable build.  Then we tag it as stable, so 
anyone can load it into a robot and start to drive.

2) Little assumptions like the joystick needs to be in slot 0, the XML 
needs to be downloaded, etc. need to be documented. Again, so anyone can 
load the robot and drive. A README or the github project wiki would be a 
good, I'll work on getting something started.

3) The dependency on an XML file downloaded separately is a problem. We 
can't have the robot code randomly crashing because the XML is missing 
or outdated. I have a couple of ideas how to resolve this, but I need to 
think about it a bit more.

Above all else though, we need stable code.  Go ahead and do 
experimental stuff on a branch, but we need a "stable" tag that really 
_is_ stable, even if it doesn't have all the bells and whistles.  We 
should _never_ be at a point now where we can't just load the robot and 
hand the stick to someone.

To this end I would like to see development done on branches, with the 
master branch reserved for merging stuff back together. We can create a 
"stable" tag from master whenever we've merged stuff and TESTED IT to 
the point we're happy with it.

## 

Steve King
steve@narbat.com mailto:steve@narbat.com

**Comment by C7C8** [comment]
The robot didn't actually crash, it just stopped giving output to the drivebase. I've been working on a theory for that - that when Parker tried to use the joystick, it just wasn't in the right slot. It's not that the code stopped working, it's that the input device it said should have existed didn't actually exist, and it has no way to check.

As for whether the stuff on the master branch is stable or not, after Parker's bug report, I was under the impression that it wasn't stable. To remedy this, I merged develop into stable, as I've been testing develop and have confirmed the code on it to be stable. Under develop, we have full control over the drivebase and intake motors. For all sakes and purposes, that code works. I wouldn't want to take it to a competition just yet - it doesn't cover things like pneumatics, the lifter, etc - but the code IS is stable enough that anybody could just load it and go.

I do try my best to confine testing code to the develop branch. The only reason I merged develop to master is to fix what I believed a stability problem (and thus render unstable code as stable).

As for updating the XML file on the robot, I too have some ideas. The code is designed such that it doesn't crash if an invalid XML file is detected (as Parker found) - another beauty of XML - but I do agree that having an entirely separate step to getting working code on the robot is a problem. Is it possible that we could modify one of Eclipse's project files to deploy the XML file in the same way it deploys the code. As far as I can tell, all Eclipse actually does is use a simple FTP file transfer to get the FRCUserProgram file into a directory; perhaps this could be adapted to work on the XML file in the same way?

For now, I'll probably just delete all the "Stable" tags.
